src/main/kotlin/com/github/avrokotlin/avro4k/serializer/AvroSerializer.kt

```diff
@@ -1,39 +1,22 @@
 package com.github.avrokotlin.avro4k.serializer
 
-import com.github.avrokotlin.avro4k.decoder.ExtendedDecoder
-import com.github.avrokotlin.avro4k.decoder.FieldDecoder
-import com.github.avrokotlin.avro4k.encoder.ExtendedEncoder
-import com.github.avrokotlin.avro4k.encoder.FieldEncoder
-import com.github.avrokotlin.avro4k.schema.extractNonNull
+import com.github.avrokotlin.avro4k.decoder.NativeAvroDecoder
+import com.github.avrokotlin.avro4k.encoder.NativeAvroEncoder
 import kotlinx.serialization.KSerializer
 import kotlinx.serialization.encoding.Decoder
 import kotlinx.serialization.encoding.Encoder
 import org.apache.avro.Schema
 
 abstract class AvroSerializer<T> : KSerializer<T> {
-
    final override fun serialize(encoder: Encoder, value: T) {
-      val schema = (encoder as FieldEncoder).fieldSchema()
-      // we may be encoding a nullable schema
-      val subschema = when (schema.type) {
-         Schema.Type.UNION -> schema.extractNonNull()
-         else -> schema
-      }
-      encodeAvroValue(subschema, encoder, value)
+      encodeAvroValue((encoder as NativeAvroEncoder).currentResolvedSchema, encoder, value)
    }
 
    final override fun deserialize(decoder: Decoder): T {
-      val schema = (decoder as FieldDecoder).fieldSchema()
-//      // we may be coming from a nullable schema aka a union
-//      val subschema = when (schema.type) {
-//         Schema.Type.UNION -> schema.extractNonNull()
-//         else -> schema
-//      }
-      return decodeAvroValue(schema, decoder)
+       return decodeAvroValue((decoder as NativeAvroDecoder).currentSchema, decoder)
    }
 
-   abstract fun encodeAvroValue(schema: Schema, encoder: ExtendedEncoder, obj: T)
+   abstract fun encodeAvroValue(schema: Schema, encoder: NativeAvroEncoder, obj: T)
```

Comment:
This is a breaking change with the outer API. If possible, I would not introduce it with this patch. There will be an upcoming interface change if we start our way to multiplatform. So let us start a discussion on how the API should look like taking the upcoming change to natively encode/decode into account. That way, we only change the API once.

Reply:
True, but there is a lot of over-interfaces with non explicit namings, so yes let's discuss about it 😄 i'll rollback the change about the naming